### PR TITLE
ARROW-9121: [C++] Forbid empty or root path in FileSystem::DeleteDirContents

### PR DESCRIPTION
**Pull request by pitrou**
Add a separate method DeleteRootDirContents, in case the operation of wiping the root directory is really desired.

### CODE REVIEW

File: cpp/src/arrow/filesystem/s3fs.cc [hunk 1]
```diff
@@ -1433,6 +1433,10 @@ Status S3FileSystem::DeleteDirContents(const std::string& s) {
   return impl_->EnsureDirectoryExists(path);
 }
 
+Status S3FileSystem::DeleteRootDirContents() {
+  return Status::NotImplemented("Cannot delete all S3 buckets");
```
Comment (kszucs) [inline cpp/src/arrow/filesystem/s3fs.cc:1437]: :)

File: python/pyarrow/_fs.pyx [hunk 2]
```diff
@@ -484,10 +484,17 @@ cdef class FileSystem:
...
         path : str
             The path of the directory to be deleted.
+        accept_root_dir : boolean, default False
+            Allow deleting the root directory's contents
+            (if path is empty or "/")
         """
         cdef c_string directory = _path_as_bytes(path)
-        with nogil:
-            check_status(self.fs.DeleteDirContents(directory))
+        if accept_root_dir and directory in (b"", b"/"):
```
Comment (kszucs) [inline python/pyarrow/_fs.pyx:492]: IsEmptyPath is true for any number of slashes, shouldn't we allow that here as well?

File: python/pyarrow/fs.py [hunk 3]
```diff
@@ -146,13 +146,22 @@ def create_dir(self, path, recursive):
...
-    def delete_dir_contents(self, path):
+    def _delete_dir_contents(self, path):
         for subpath in self.fs.listdir(path, detail=False):
             if self.fs.isdir(subpath):
                 self.fs.rm(subpath, recursive=True)
             elif self.fs.isfile(subpath):
                 self.fs.rm(subpath)
 
+    def delete_dir_contents(self, path):
+        if path in ("", "/"):
```
Comment (kszucs) [inline python/pyarrow/fs.py:157]: Same as above, do we want to allow multiple slashes?

File: python/pyarrow/_fs.pyx (hunk 2 above)
Comment (pitrou) [reply python/pyarrow/_fs.pyx:492]: Sure.

**Comment by kszucs** [comment]
Thanks for taking it over. It looks good to me except handling the multiple slashes on the python side.

**Review by kszucs — APPROVED** [review]
Thanks Antoine!

**Pull request closed without merging**